graphblas_algorithms/algorithms/shortest_paths/weighted.py

```diff
@@ -0,0 +1,45 @@
+from graphblas import Vector, binary, monoid, replace
+from graphblas.semiring import min_plus
+
+from ..exceptions import Unbounded
+
+__all__ = ["single_source_bellman_ford_path_length"]
+
+
+def single_source_bellman_ford_path_length(G, source):
+    # No need for `is_weighted=` keyword, b/c this is assumed to be weighted (I think)
+    index = G._key_to_id[source]
+    A = G._A
+    if A.dtype == bool:
+        # Should we upcast e.g. INT8 to INT64 as well?
+        dtype = int
+    else:
+        dtype = A.dtype
+    n = A.nrows
+    d = Vector(dtype, n, name="single_source_bellman_ford_path_length")
+    d[index] = 0
+    cur = d.dup(name="cur")
+    mask = Vector(bool, n, name="mask")
+    for _i in range(n - 1):
+        # This is a slightly modified Bellman-Ford algorithm.
+        # `cur` is the current frontier of values that improved in the previous iteration.
+        # This means that in this iteration we drop values from `cur` that are not better.
+        cur << min_plus(cur @ A)
+
+        # Mask is True where cur not in d or cur < d
+        mask(cur.S, replace) << True  # or: `mask << unary.one[bool](cur)`
+        mask(binary.second) << binary.lt(cur & d)
+
+        # Drop values from `cur` that didn't improve
+        cur(mask.V, replace) << cur
```

Comment:
FYI, a very similar pattern to this also showed up in Floyd-Warshall
https://github.com/python-graphblas/graphblas-algorithms/blob/0b649b2556e6c21f547f24e39c63bb206eb22f04/graphblas_algorithms/algorithms/shortest_paths/dense.py#L67-L73
where we need to set a mask according to a comparison _and_ set the mask to True for new values.